spark/src/main/scala/io/substrait/spark/ToSubstraitType.scala

```diff
@@ -76,13 +83,47 @@ private class ToSparkType
       expr.key().accept(this),
       expr.value().accept(this),
       valueContainsNull = expr.value().nullable())
+
+  override def visit(expr: Type.Struct): DataType =
+    StructType(
+      expr.fields.asScala.zipWithIndex
+        .map {
+          case (f, i) =>
+            val name = if (dfsNames.nonEmpty) {
+              require(nameIdx < dfsNames.size)
+              val n = dfsNames(nameIdx)
+              nameIdx += 1
+              n
```

Comment:
```suggestion
              dfsNames(i)
```
Won't `nameIdx` always be the same `i`? If so, no need to increment a separate counter.
Incrementing an object-scoped counter assumes that the `.map` is behaving like a `foreach` - i.e. guaranteed order of execution, but I think (in theory, at least) the map could execute its functions concurrently (not sure it does, in practice), just assembling the results back into the original sequence.

Reply:
Looking at this further, I guess it's doing this because it's doing a depth-first traversal of a potentially nested struct.

Reply:
Yup, exactly